vm/devices/net/netvsp/src/lib.rs

```diff
@@ -5519,28 +5520,32 @@ impl<T: 'static + RingMem> NetChannel<T> {
             .map_err(WorkerError::GpaDirectError)?;
 
         let mut reader = packet.rndis_reader(&buffers.mem);
-        // There may be multiple RNDIS packets in a single
-        // message, concatenated with each other. Consume
-        // them until there is no more data in the RNDIS
-        // message.
-        while reader.len() > 0 {
-            let mut this_reader = reader.clone();
-            let header: rndisprot::MessageHeader = this_reader.read_plain()?;
-            if self.handle_rndis_message(
+        let header: rndisprot::MessageHeader = reader.read_plain()?;
+        if header.message_type == rndisprot::MESSAGE_TYPE_PACKET_MSG {
+            let start = segments.len();
+            match self.handle_rndis_packet_messages(
                 buffers,
                 state,
                 id,
-                header.message_type,
-                this_reader,
+                header.message_length as usize,
+                reader,
                 segments,
-                pending_control_messages,
-            )? {
-                num_packets += 1;
+            ) {
+                Ok(n) => {
+                    state.pending_tx_packets[id.0 as usize].pending_packet_count += n;
+                    total_packets += n;
+                }
+                Err(err) => {
+                    // Roll back any segments added for this message.
+                    segments.truncate(start);
+                    return Err(err);
+                }
             }
-            reader.skip(header.message_length as usize)?;
+        } else {
+            self.handle_rndis_message(state, header.message_type, reader)?;
```

Comment:
for my understanding, can the packet not have multiple non-packet RNDIS messages?

Reply:
Right.